### DISCUSSION TITLE
[BUG] JSQLParser Version 4.7  : WithItems moved inside Select

### DISCUSSION THREAD

**Discussion by serhiizapalskyi** [JSQLParser/JSqlParser#1896] [Q&A]
### Failing SQL Feature:

I'm not sure is it a bug or feature, but unfortunately it breaks our current behavior. After migration from 4.5 to 4.7 the query `WITH events as (select * from table1) select t1 from  events`  is parsed with select body including WithItems. In previous versions Select.getSelectBody() was returning select without WithItems.

So Select.getSelectBody  (I know it is deprecated) will result different in 4.5 and 4.7:
4.5: `select t1 from  events`
4.7 `WITH events as (select * from table1) select t1 from  events`

I see that Select structure was changed: 
In 4.5 there was separately `selectBody` and `withItems`
In 4.7 `withItems` are included into `select`

#### The question: 
Can I get select without WithItems (like it was in 4.5)?

### SQL Example:
- Simplified Query Example, focusing on the failing feature
    ```
          WITH events as (
                 select * from table1
           ) 
           select t1 
           from  events
    ```

### Software Information:
- JSqlParser version 4.7

### Tips:
The same query has different structure in different versions of parser

4.5 Structure:

<img width="874" alt="Screenshot 2023-11-22 at 13 44 45" src="https://github.com/JSQLParser/JSqlParser/assets/148860688/b6edecfc-c63e-495f-beb7-45743476e1ff">

4.7 Structure:

<img width="740" alt="image" src="https://github.com/JSQLParser/JSqlParser/assets/148860688/a1590a6f-ea94-4a44-86aa-da6287c046b3">

Thanks in advance,
Regards, Serhii Zapalskyi

**Comment by manticore-projects** [comment]
Greetings!

Thank you for your question. Although I have read it carefully three time and I am still not sure what your are asking for.
With JSQLParser-4.7 your AST will look like shown below. [Your can try this by yourself online and interactively.](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20OoSQKgEgBApgbjAdgFwM5QIIGUoAoBQURxRWAogDJkDCYUAVIScQGIBKA8gLJTICGAIwA2MAIxQAlPnJVavUfnbdYCFKnwBuIA)

Are you trying to to eliminate the `With` clause from the `PlainSelect`? If so, then why not just ignore the `PlainSelect`'s property `withItemsList` or set it to `NULL` (in case your are rewriting).


![image](https://github.com/JSQLParser/JSqlParser/assets/18080123/be0117c1-3b9a-41bf-9b61-5759cce9498b)

**Comment by manticore-projects** [comment]
Here is the Java API for `PlainSelect` showing how to modify the `withItemsList` property:
https://manticore-projects.com/JSQLParser/javadoc_stable.html#select

**Comment by serhiizapalskyi** [comment]
What I want, to use `select`  like it was in previous version:
1. get whole select body without withItems from initial select (`select t1  from  events`)
2. hold initial select unchanged:
  ```
WITH events as (
             select * from table1
       ) 
       select t1 
       from  events
```
3. It would be great to understand why it has changed. 
Before 4.7: 

<img width="869" alt="Screenshot 2023-11-27 at 09 43 44" src="https://github.com/JSQLParser/JSqlParser/assets/148860688/39f50f47-62a2-4b37-a6b5-9fbf002f7411">

<img width="868" alt="Screenshot 2023-11-27 at 09 45 27" src="https://github.com/JSQLParser/JSqlParser/assets/148860688/bbed6249-17a0-4d46-9004-948907c10d59">

After:

<img width="912" alt="Screenshot 2023-11-27 at 09 37 10" src="https://github.com/JSQLParser/JSqlParser/assets/148860688/b71ef62a-1d91-40a0-abc2-1fd19aaacd74">

**Comment by serhiizapalskyi** [comment]
And I have one more question about this bug https://github.com/JSQLParser/JSqlParser/issues/1856

I'm migrating from 4.0 to 4.7 and IGNORE NULLS in ARRAY_AGG parsed successfully in 4.2, but does not in 4.3. Can you confirm that it is the same issue which is described in [here](https://github.com/JSQLParser/JSqlParser/issues/1856)  ?

This is the failed query

`select ARRAY_AGG(DISTINCT NULLIF(wm_uservars_name, '') IGNORE NULLS) from ecs_db.prod_db_sessions_1`

**Comment by manticore-projects** [comment]
> What I want, to use `select` like it was in previous version:
> 
>     1. get whole select body without withItems from initial select (`select t1  from  events`)
> 
>     2. hold initial select unchanged:

If you are interested in the `AST`, then access the `PlainSelect` and just ignore its property `WithItemsList`.
If you are interested in deparsing the `AST`, then either set the property `WithItemsList` to `NULL` or overwrite the `WithItem` Visitor.

>  3. It would be great to understand why it has changed.

For various reasons, but the most obvious is the fact that a) the `With` clause belongs to a statement and can't exist on its own and b) this statement depends on the `With` clause and won't execute without it (unless the `With` is redundant). So `WithItemsList` is a natural property of a `Statement`.

**Comment by manticore-projects** [comment]
> Can you confirm that it is the same issue which is described in [here](https://github.com/JSQLParser/JSqlParser/issues/1856) ?

Very likely, `ARRAY_AGG` is a special function with a special parameter syntax which has not been implemented.
You would need to provide or sponsor an implementation when you are interested in this particular feature.

**Comment by serhiizapalskyi** [comment]
Thank you for the answers
